{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.7.6 64-bit ('pbt')",
   "metadata": {
    "interpreter": {
     "hash": "ec5ba68ec718e9e4ee669959602a0353a6af89966bd0f0090d74b47bfbf3a129"
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "#from sklearn import train_test_split\n",
    "import sklearn.metrics\n",
    "import dataclasses\n",
    "import lightgbm as lgb\n",
    "from ray import tune\n",
    "from ray.tune.schedulers import PopulationBasedTraining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_processed_df = pd.read_pickle(\"inputs/train_processed.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search_space = {\n",
    "\n",
    "# }\n",
    "\n",
    "# pbt_scheduler = PopulationBasedTraining(\n",
    "#         time_attr='time_total_s',\n",
    "#         metric='mean_accuracy',\n",
    "#         mode='max',\n",
    "#         perturbation_interval=600.0,\n",
    "#         hyperparam_mutations={\n",
    "#             \"lr\": [1e-3, 5e-4, 1e-4, 5e-5, 1e-5],\n",
    "#             \"alpha\": lambda: random.uniform(0.0, 1.0),\n",
    "#         ...\n",
    "#         })\n",
    "# tune.run( train_processed_df, scheduler=pbt_scheduler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sklearn.model_selection import train_test_split\n",
    "# X=train_processed_df.drop('diabetes_mellitus', axis=1)\n",
    "# y=train_processed_df.diabetes_mellitus\n",
    "# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=314, stratify=y)\n",
    "# train_set = lgb.Dataset(X_train, label=y_train)\n",
    "# test_set = lgb.Dataset(X_test, label=y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def LightGBMCallback(env):\n",
    "    \"\"\"Assumes that `valid_0` is the target validation score.\"\"\"\n",
    "    _, metric, score, _ = env.evaluation_result_list[0]\n",
    "    tune.report(**{metric: score})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_diabetes(config):\n",
    "    #data, target = sklearn.datasets.load_breast_cancer(return_X_y=True)\n",
    "    #train_processed_df = pd.read_pickle(\"inputs/train_processed.pkl\")\n",
    "    data = train_processed_df.drop('diabetes_mellitus', axis=1)\n",
    "    target = train_processed_df.diabetes_mellitus\n",
    "    train_x, test_x, train_y, test_y = train_test_split(\n",
    "        data, target, test_size=0.25)\n",
    "    train_set = lgb.Dataset(train_x, label=train_y)\n",
    "    test_set = lgb.Dataset(test_x, label=test_y)\n",
    "    gbm = lgb.train(\n",
    "        config,\n",
    "        train_set,\n",
    "        valid_sets=[test_set],\n",
    "        verbose_eval=False,\n",
    "        callbacks=[LightGBMCallback])\n",
    "    preds = gbm.predict(test_x)\n",
    "    pred_labels = np.rint(preds)\n",
    "    tune.report(\n",
    "        mean_accuracy=sklearn.metrics.roc_auc_score(test_y, pred_labels),\n",
    "        done=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "tags": [
     "outputPrepend"
    ]
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-03-18 15:03:27,518\tWARNING worker.py:1107 -- Warning: The actor ImplicitFunc has size 85149703 when pickled. It will be stored in Redis, which could cause memory issues. This may mean that its definition uses a large array or other object.\n",
      "2021-03-18 15:03:27,646\tWARNING util.py:152 -- The `start_trial` operation took 0.858 s, which may be a performance bottleneck.\n"
     ]
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "== Status ==<br>Memory usage on this node: 8.1/16.0 GiB<br>PopulationBasedTraining: 0 checkpoints, 0 perturbs<br>Resources requested: 1/8 CPUs, 0/0 GPUs, 0.0/4.59 GiB heap, 0.0/1.56 GiB objects<br>Result logdir: /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25<br>Number of trials: 1/4 (1 RUNNING)<br><table>\n<thead>\n<tr><th>Trial name                </th><th>status  </th><th>loc  </th><th>boosting_type  </th><th style=\"text-align: right;\">  learning_rate</th><th style=\"text-align: right;\">  num_leaves</th></tr>\n</thead>\n<tbody>\n<tr><td>train_diabetes_17355_00000</td><td>RUNNING </td><td>     </td><td>gbdt           </td><td style=\"text-align: right;\">      0.0200876</td><td style=\"text-align: right;\">         439</td></tr>\n</tbody>\n</table><br><br>"
     },
     "metadata": {}
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-03-18 15:03:28,892\tWARNING util.py:152 -- The `start_trial` operation took 0.679 s, which may be a performance bottleneck.\n",
      "2021-03-18 15:03:30,083\tWARNING util.py:152 -- The `start_trial` operation took 0.733 s, which may be a performance bottleneck.\n",
      "2021-03-18 15:03:31,472\tWARNING util.py:152 -- The `start_trial` operation took 0.765 s, which may be a performance bottleneck.\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/basic.py:1286: UserWarning: Overriding the parameters from Reference Dataset.\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   warnings.warn('Overriding the parameters from Reference Dataset.')\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/basic.py:1098: UserWarning: categorical_column in param dict is overridden.\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   warnings.warn('{} in param dict is overridden.'.format(cat_alias))\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/basic.py:1286: UserWarning: Overriding the parameters from Reference Dataset.\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   warnings.warn('Overriding the parameters from Reference Dataset.')\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m 2021-03-18 15:03:42,146\tERROR function_runner.py:254 -- Runner Thread raised error.\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     self._entrypoint()\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     self._status_reporter.get_checkpoint())\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m Exception in thread Thread-2:\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/opt/anaconda3/lib/python3.7/threading.py\", line 926, in _bootstrap_inner\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     self.run()\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 267, in run\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     raise e\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     self._entrypoint()\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     self._status_reporter.get_checkpoint())\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35848)\u001b[0m \n",
      "2021-03-18 15:03:42,317\tERROR trial_runner.py:616 -- Trial train_diabetes_17355_00000: Error processing event.\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trial_runner.py\", line 586, in _process_trial\n",
      "    results = self.trial_executor.fetch_result(trial)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py\", line 609, in fetch_result\n",
      "    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/_private/client_mode_hook.py\", line 47, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/worker.py\", line 1456, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(TuneError): \u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35848, ip=192.168.0.4)\n",
      "  File \"python/ray/_raylet.pyx\", line 480, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 432, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 167, in train_buffered\n",
      "    result = self.train()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 226, in train\n",
      "    result = self.step()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 366, in step\n",
      "    self._report_thread_runner_error(block=True)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 513, in _report_thread_runner_error\n",
      "    (\"Trial raised an exception. Traceback:\\n{}\".format(err_tb_str)\n",
      "ray.tune.error.TuneError: Trial raised an exception. Traceback:\n",
      "\u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35848, ip=192.168.0.4)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "    self._entrypoint()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "    self._status_reporter.get_checkpoint())\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "    evaluation_result_list=evaluation_result_list))\n",
      "  File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "IndexError: list index out of range\n",
      "Result for train_diabetes_17355_00000:\n",
      "  {}\n",
      "  \n"
     ]
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "== Status ==<br>Memory usage on this node: 8.5/16.0 GiB<br>PopulationBasedTraining: 0 checkpoints, 0 perturbs<br>Resources requested: 3/8 CPUs, 0/0 GPUs, 0.0/4.59 GiB heap, 0.0/1.56 GiB objects<br>Result logdir: /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25<br>Number of trials: 4/4 (1 ERROR, 3 RUNNING)<br><table>\n<thead>\n<tr><th>Trial name                </th><th>status  </th><th>loc  </th><th>boosting_type  </th><th style=\"text-align: right;\">  learning_rate</th><th style=\"text-align: right;\">  num_leaves</th></tr>\n</thead>\n<tbody>\n<tr><td>train_diabetes_17355_00001</td><td>RUNNING </td><td>     </td><td>dart           </td><td style=\"text-align: right;\">      0.0738899</td><td style=\"text-align: right;\">         220</td></tr>\n<tr><td>train_diabetes_17355_00002</td><td>RUNNING </td><td>     </td><td>gbdt           </td><td style=\"text-align: right;\">      0.0420212</td><td style=\"text-align: right;\">         493</td></tr>\n<tr><td>train_diabetes_17355_00003</td><td>RUNNING </td><td>     </td><td>dart           </td><td style=\"text-align: right;\">      0.064954 </td><td style=\"text-align: right;\">         953</td></tr>\n<tr><td>train_diabetes_17355_00000</td><td>ERROR   </td><td>     </td><td>gbdt           </td><td style=\"text-align: right;\">      0.0200876</td><td style=\"text-align: right;\">         439</td></tr>\n</tbody>\n</table><br>Number of errored trials: 1<br><table>\n<thead>\n<tr><th>Trial name                </th><th style=\"text-align: right;\">  # failures</th><th>error file                                                                                                                                                                                                                                     </th></tr>\n</thead>\n<tbody>\n<tr><td>train_diabetes_17355_00000</td><td style=\"text-align: right;\">           1</td><td>/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25/train_diabetes_17355_00000_0_boosting_type=gbdt,learning_rate=0.020088,num_leaves=439_2021-03-18_15-03-26/error.txt</td></tr>\n</tbody>\n</table><br>"
     },
     "metadata": {}
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "nt/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m Exception in thread Thread-2:\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"/Users/anitaclement/opt/anaconda3/lib/python3.7/threading.py\", line 926, in _bootstrap_inner\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     self.run()\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 267, in run\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     raise e\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     self._entrypoint()\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     self._status_reporter.get_checkpoint())\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35863)\u001b[0m \n",
      "2021-03-18 15:03:43,839\tERROR trial_runner.py:616 -- Trial train_diabetes_17355_00001: Error processing event.\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trial_runner.py\", line 586, in _process_trial\n",
      "    results = self.trial_executor.fetch_result(trial)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py\", line 609, in fetch_result\n",
      "    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/_private/client_mode_hook.py\", line 47, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/worker.py\", line 1456, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(TuneError): \u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35863, ip=192.168.0.4)\n",
      "  File \"python/ray/_raylet.pyx\", line 480, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 432, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 167, in train_buffered\n",
      "    result = self.train()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 226, in train\n",
      "    result = self.step()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 366, in step\n",
      "    self._report_thread_runner_error(block=True)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 513, in _report_thread_runner_error\n",
      "    (\"Trial raised an exception. Traceback:\\n{}\".format(err_tb_str)\n",
      "ray.tune.error.TuneError: Trial raised an exception. Traceback:\n",
      "\u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35863, ip=192.168.0.4)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "    self._entrypoint()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "    self._status_reporter.get_checkpoint())\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "    evaluation_result_list=evaluation_result_list))\n",
      "  File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "IndexError: list index out of range\n",
      "Result for train_diabetes_17355_00001:\n",
      "  {}\n",
      "  \n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/basic.py:1098: UserWarning: categorical_column in param dict is overridden.\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   warnings.warn('{} in param dict is overridden.'.format(cat_alias))\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/basic.py:1286: UserWarning: Overriding the parameters from Reference Dataset.\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   warnings.warn('Overriding the parameters from Reference Dataset.')\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/basic.py:1098: UserWarning: categorical_column in param dict is overridden.\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   warnings.warn('{} in param dict is overridden.'.format(cat_alias))\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m 2021-03-18 15:03:45,449\tERROR function_runner.py:254 -- Runner Thread raised error.\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     self._entrypoint()\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     self._status_reporter.get_checkpoint())\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m Exception in thread Thread-2:\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/opt/anaconda3/lib/python3.7/threading.py\", line 926, in _bootstrap_inner\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     self.run()\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 267, in run\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     raise e\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     self._entrypoint()\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     self._status_reporter.get_checkpoint())\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35866)\u001b[0m \n",
      "2021-03-18 15:03:45,623\tERROR trial_runner.py:616 -- Trial train_diabetes_17355_00002: Error processing event.\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trial_runner.py\", line 586, in _process_trial\n",
      "    results = self.trial_executor.fetch_result(trial)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py\", line 609, in fetch_result\n",
      "    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/_private/client_mode_hook.py\", line 47, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/worker.py\", line 1456, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(TuneError): \u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35866, ip=192.168.0.4)\n",
      "  File \"python/ray/_raylet.pyx\", line 480, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 432, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 167, in train_buffered\n",
      "    result = self.train()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 226, in train\n",
      "    result = self.step()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 366, in step\n",
      "    self._report_thread_runner_error(block=True)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 513, in _report_thread_runner_error\n",
      "    (\"Trial raised an exception. Traceback:\\n{}\".format(err_tb_str)\n",
      "ray.tune.error.TuneError: Trial raised an exception. Traceback:\n",
      "\u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35866, ip=192.168.0.4)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "    self._entrypoint()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "    self._status_reporter.get_checkpoint())\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "    evaluation_result_list=evaluation_result_list))\n",
      "  File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "IndexError: list index out of range\n",
      "Result for train_diabetes_17355_00002:\n",
      "  {}\n",
      "  \n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m 2021-03-18 15:03:46,893\tERROR function_runner.py:254 -- Runner Thread raised error.\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     self._entrypoint()\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     self._status_reporter.get_checkpoint())\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m Exception in thread Thread-2:\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m Traceback (most recent call last):\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/opt/anaconda3/lib/python3.7/threading.py\", line 926, in _bootstrap_inner\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     self.run()\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 267, in run\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     raise e\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     self._entrypoint()\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     self._status_reporter.get_checkpoint())\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     output = fn()\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m     evaluation_result_list=evaluation_result_list))\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m   File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m IndexError: list index out of range\n",
      "\u001b[2m\u001b[36m(pid=35869)\u001b[0m \n",
      "2021-03-18 15:03:47,177\tERROR trial_runner.py:616 -- Trial train_diabetes_17355_00003: Error processing event.\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trial_runner.py\", line 586, in _process_trial\n",
      "    results = self.trial_executor.fetch_result(trial)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py\", line 609, in fetch_result\n",
      "    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/_private/client_mode_hook.py\", line 47, in wrapper\n",
      "    return func(*args, **kwargs)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/worker.py\", line 1456, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(TuneError): \u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35869, ip=192.168.0.4)\n",
      "  File \"python/ray/_raylet.pyx\", line 480, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 432, in ray._raylet.execute_task.function_executor\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 167, in train_buffered\n",
      "    result = self.train()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trainable.py\", line 226, in train\n",
      "    result = self.step()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 366, in step\n",
      "    self._report_thread_runner_error(block=True)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 513, in _report_thread_runner_error\n",
      "    (\"Trial raised an exception. Traceback:\\n{}\".format(err_tb_str)\n",
      "ray.tune.error.TuneError: Trial raised an exception. Traceback:\n",
      "\u001b[36mray::ImplicitFunc.train_buffered()\u001b[39m (pid=35869, ip=192.168.0.4)\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 248, in run\n",
      "    self._entrypoint()\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 316, in entrypoint\n",
      "    self._status_reporter.get_checkpoint())\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/function_runner.py\", line 576, in _trainable_func\n",
      "    output = fn()\n",
      "  File \"<ipython-input-35-e67547cfe31f>\", line 15, in train_diabetes\n",
      "  File \"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/lightgbm/engine.py\", line 267, in train\n",
      "    evaluation_result_list=evaluation_result_list))\n",
      "  File \"<ipython-input-17-c4e673d45b99>\", line 3, in LightGBMCallback\n",
      "IndexError: list index out of range\n",
      "Result for train_diabetes_17355_00003:\n",
      "  {}\n",
      "  \n"
     ]
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "== Status ==<br>Memory usage on this node: 7.5/16.0 GiB<br>PopulationBasedTraining: 0 checkpoints, 0 perturbs<br>Resources requested: 0/8 CPUs, 0/0 GPUs, 0.0/4.59 GiB heap, 0.0/1.56 GiB objects<br>Result logdir: /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25<br>Number of trials: 4/4 (4 ERROR)<br><table>\n<thead>\n<tr><th>Trial name                </th><th>status  </th><th>loc  </th><th>boosting_type  </th><th style=\"text-align: right;\">  learning_rate</th><th style=\"text-align: right;\">  num_leaves</th></tr>\n</thead>\n<tbody>\n<tr><td>train_diabetes_17355_00000</td><td>ERROR   </td><td>     </td><td>gbdt           </td><td style=\"text-align: right;\">      0.0200876</td><td style=\"text-align: right;\">         439</td></tr>\n<tr><td>train_diabetes_17355_00001</td><td>ERROR   </td><td>     </td><td>dart           </td><td style=\"text-align: right;\">      0.0738899</td><td style=\"text-align: right;\">         220</td></tr>\n<tr><td>train_diabetes_17355_00002</td><td>ERROR   </td><td>     </td><td>gbdt           </td><td style=\"text-align: right;\">      0.0420212</td><td style=\"text-align: right;\">         493</td></tr>\n<tr><td>train_diabetes_17355_00003</td><td>ERROR   </td><td>     </td><td>dart           </td><td style=\"text-align: right;\">      0.064954 </td><td style=\"text-align: right;\">         953</td></tr>\n</tbody>\n</table><br>Number of errored trials: 4<br><table>\n<thead>\n<tr><th>Trial name                </th><th style=\"text-align: right;\">  # failures</th><th>error file                                                                                                                                                                                                                                     </th></tr>\n</thead>\n<tbody>\n<tr><td>train_diabetes_17355_00000</td><td style=\"text-align: right;\">           1</td><td>/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25/train_diabetes_17355_00000_0_boosting_type=gbdt,learning_rate=0.020088,num_leaves=439_2021-03-18_15-03-26/error.txt</td></tr>\n<tr><td>train_diabetes_17355_00001</td><td style=\"text-align: right;\">           1</td><td>/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25/train_diabetes_17355_00001_1_boosting_type=dart,learning_rate=0.07389,num_leaves=220_2021-03-18_15-03-28/error.txt </td></tr>\n<tr><td>train_diabetes_17355_00002</td><td style=\"text-align: right;\">           1</td><td>/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25/train_diabetes_17355_00002_2_boosting_type=gbdt,learning_rate=0.042021,num_leaves=493_2021-03-18_15-03-29/error.txt</td></tr>\n<tr><td>train_diabetes_17355_00003</td><td style=\"text-align: right;\">           1</td><td>/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_15-03-25/train_diabetes_17355_00003_3_boosting_type=dart,learning_rate=0.064954,num_leaves=953_2021-03-18_15-03-30/error.txt</td></tr>\n</tbody>\n</table><br>"
     },
     "metadata": {}
    },
    {
     "output_type": "error",
     "ename": "TuneError",
     "evalue": "('Trials did not complete', [train_diabetes_17355_00000, train_diabetes_17355_00001, train_diabetes_17355_00002, train_diabetes_17355_00003])",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTuneError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-e839d9eeaa76>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     16\u001b[0m         \u001b[0;31m#scheduler=ASHAScheduler()\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m         \u001b[0mlocal_dir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m         \u001b[0mscheduler\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mPopulationBasedTraining\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m         )\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/tune.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(run_or_experiment, name, metric, mode, stop, time_budget_s, config, resources_per_trial, num_samples, local_dir, search_alg, scheduler, keep_checkpoints_num, checkpoint_score_attr, checkpoint_freq, checkpoint_at_end, verbose, progress_reporter, log_to_file, trial_name_creator, trial_dirname_creator, sync_config, export_formats, max_failures, fail_fast, restore, server_port, resume, queue_trials, reuse_actors, trial_executor, raise_on_failed_trial, callbacks, loggers, ray_auto_init, run_errored_only, global_checkpoint_period, with_server, upload_dir, sync_to_cloud, sync_to_driver, sync_on_checkpoint)\u001b[0m\n\u001b[1;32m    442\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mincomplete_trials\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    443\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mraise_on_failed_trial\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 444\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mTuneError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Trials did not complete\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mincomplete_trials\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    445\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    446\u001b[0m             \u001b[0mlogger\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Trials did not complete: %s\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mincomplete_trials\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTuneError\u001b[0m: ('Trials did not complete', [train_diabetes_17355_00000, train_diabetes_17355_00001, train_diabetes_17355_00002, train_diabetes_17355_00003])"
     ]
    }
   ],
   "source": [
    "# config = {\n",
    "#         \"objective\": \"binary\",\n",
    "#         #\"metric\": \"binary_error\",\n",
    "#         \"metric\": \"mean_accuracy\",\n",
    "#         \"verbose\": -1,\n",
    "#         \"boosting_type\": tune.grid_search([\"gbdt\", \"dart\"]),\n",
    "#         \"num_leaves\": tune.randint(10, 1000),\n",
    "#         \"learning_rate\": tune.loguniform(1e-2, 1e-1)\n",
    "#     }\n",
    "# analysis = tune.run(\n",
    "#         train_diabetes,\n",
    "#         metric=\"auc\",\n",
    "#         mode=\"max\",\n",
    "#         config=config,\n",
    "#         num_samples=2, #Number of times to sample from the hyperparameter space\n",
    "#         #scheduler=ASHAScheduler()\n",
    "#         local_dir=\"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging\",\n",
    "#         scheduler=PopulationBasedTraining()\n",
    "#         )\n",
    "\n",
    "# print(\"Best hyperparameters found were: \", analysis.best_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-03-18 16:08:41,721\tWARNING worker.py:1107 -- Warning: The actor ImplicitFunc has size 85149680 when pickled. It will be stored in Redis, which could cause memory issues. This may mean that its definition uses a large array or other object.\n",
      "2021-03-18 16:08:41,862\tWARNING util.py:152 -- The `start_trial` operation took 0.724 s, which may be a performance bottleneck.\n"
     ]
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "== Status ==<br>Memory usage on this node: 9.0/16.0 GiB<br>PopulationBasedTraining: 0 checkpoints, 0 perturbs<br>Resources requested: 1/8 CPUs, 0/0 GPUs, 0.0/4.59 GiB heap, 0.0/1.56 GiB objects<br>Result logdir: /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_16-08-39<br>Number of trials: 1/1764 (1 RUNNING)<br><table>\n<thead>\n<tr><th>Trial name                </th><th>status  </th><th>loc  </th><th style=\"text-align: right;\">  colsample_bytree</th><th style=\"text-align: right;\">  min_child_samples</th><th style=\"text-align: right;\">  min_child_weight</th><th style=\"text-align: right;\">  num_leaves</th><th style=\"text-align: right;\">  reg_alpha</th><th style=\"text-align: right;\">  reg_lambda</th><th style=\"text-align: right;\">  scale_pos_weight</th><th style=\"text-align: right;\">  subsample</th></tr>\n</thead>\n<tbody>\n<tr><td>train_diabetes_344c8_00000</td><td>RUNNING </td><td>     </td><td style=\"text-align: right;\">          0.283789</td><td style=\"text-align: right;\">                122</td><td style=\"text-align: right;\">             1e-07</td><td style=\"text-align: right;\">          36</td><td style=\"text-align: right;\">          0</td><td style=\"text-align: right;\">           0</td><td style=\"text-align: right;\">                 1</td><td style=\"text-align: right;\">    0.46226</td></tr>\n</tbody>\n</table><br><br>"
     },
     "metadata": {}
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-03-18 16:08:42,848\tWARNING util.py:152 -- The `start_trial` operation took 0.520 s, which may be a performance bottleneck.\n",
      "2021-03-18 16:08:43,902\tWARNING util.py:152 -- The `start_trial` operation took 0.630 s, which may be a performance bottleneck.\n",
      "2021-03-18 16:08:45,154\tWARNING util.py:152 -- The `start_trial` operation took 0.691 s, which may be a performance bottleneck.\n",
      "2021-03-18 16:08:46,266\tWARNING util.py:152 -- The `start_trial` operation took 0.590 s, which may be a performance bottleneck.\n",
      "2021-03-18 16:08:47,293\tWARNING util.py:152 -- The `start_trial` operation took 0.599 s, which may be a performance bottleneck.\n"
     ]
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "== Status ==<br>Memory usage on this node: 9.4/16.0 GiB<br>PopulationBasedTraining: 0 checkpoints, 0 perturbs<br>Resources requested: 6/8 CPUs, 0/0 GPUs, 0.0/4.59 GiB heap, 0.0/1.56 GiB objects<br>Result logdir: /Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging/train_diabetes_2021-03-18_16-08-39<br>Number of trials: 6/1764 (6 RUNNING)<br><table>\n<thead>\n<tr><th>Trial name                </th><th>status  </th><th>loc  </th><th style=\"text-align: right;\">  colsample_bytree</th><th style=\"text-align: right;\">  min_child_samples</th><th style=\"text-align: right;\">  min_child_weight</th><th style=\"text-align: right;\">  num_leaves</th><th style=\"text-align: right;\">  reg_alpha</th><th style=\"text-align: right;\">  reg_lambda</th><th style=\"text-align: right;\">  scale_pos_weight</th><th style=\"text-align: right;\">  subsample</th></tr>\n</thead>\n<tbody>\n<tr><td>train_diabetes_344c8_00000</td><td>RUNNING </td><td>     </td><td style=\"text-align: right;\">          0.283789</td><td style=\"text-align: right;\">                122</td><td style=\"text-align: right;\">            1e-07 </td><td style=\"text-align: right;\">          36</td><td style=\"text-align: right;\">          0</td><td style=\"text-align: right;\">           0</td><td style=\"text-align: right;\">                 1</td><td style=\"text-align: right;\">   0.46226 </td></tr>\n<tr><td>train_diabetes_344c8_00001</td><td>RUNNING </td><td>     </td><td style=\"text-align: right;\">          0.430756</td><td style=\"text-align: right;\">                288</td><td style=\"text-align: right;\">            1e-06 </td><td style=\"text-align: right;\">          16</td><td style=\"text-align: right;\">          0</td><td style=\"text-align: right;\">           0</td><td style=\"text-align: right;\">                 1</td><td style=\"text-align: right;\">   0.435398</td></tr>\n<tr><td>train_diabetes_344c8_00002</td><td>RUNNING </td><td>     </td><td style=\"text-align: right;\">          0.318188</td><td style=\"text-align: right;\">                279</td><td style=\"text-align: right;\">            1e-05 </td><td style=\"text-align: right;\">          38</td><td style=\"text-align: right;\">          0</td><td style=\"text-align: right;\">           0</td><td style=\"text-align: right;\">                 1</td><td style=\"text-align: right;\">   0.482496</td></tr>\n<tr><td>train_diabetes_344c8_00003</td><td>RUNNING </td><td>     </td><td style=\"text-align: right;\">          0.512273</td><td style=\"text-align: right;\">                229</td><td style=\"text-align: right;\">            0.0001</td><td style=\"text-align: right;\">          16</td><td style=\"text-align: right;\">          0</td><td style=\"text-align: right;\">           0</td><td style=\"text-align: right;\">                 1</td><td style=\"text-align: right;\">   0.589058</td></tr>\n<tr><td>train_diabetes_344c8_00004</td><td>RUNNING </td><td>     </td><td style=\"text-align: right;\">          0.347268</td><td style=\"text-align: right;\">                126</td><td style=\"text-align: right;\">            0.001 </td><td style=\"text-align: right;\">          27</td><td style=\"text-align: right;\">          0</td><td style=\"text-align: right;\">           0</td><td style=\"text-align: right;\">                 1</td><td style=\"text-align: right;\">   0.567391</td></tr>\n<tr><td>train_diabetes_344c8_00005</td><td>RUNNING </td><td>     </td><td style=\"text-align: right;\">          0.247092</td><td style=\"text-align: right;\">                291</td><td style=\"text-align: right;\">            0.01  </td><td style=\"text-align: right;\">          29</td><td style=\"text-align: right;\">          0</td><td style=\"text-align: right;\">           0</td><td style=\"text-align: right;\">                 1</td><td style=\"text-align: right;\">   0.427536</td></tr>\n</tbody>\n</table><br><br>"
     },
     "metadata": {}
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-03-18 16:08:48,221\tWARNING util.py:152 -- The `start_trial` operation took 0.538 s, which may be a performance bottleneck.\n",
      "2021-03-18 16:08:49,186\tWARNING util.py:152 -- The `start_trial` operation took 0.543 s, which may be a performance bottleneck.\n",
      "2021-03-18 16:08:54,267\tWARNING worker.py:1107 -- The actor or task with ID ffffffffffffffffd398a33839d669fb961ecf3d01000000 cannot be scheduled right now. It requires {CPU: 1.000000} for placement, but this node only has remaining {0.000000/8.000000 CPU, 4.589844 GiB/4.589844 GiB memory, 1.000000/1.000000 node:192.168.0.4, 1.562500 GiB/1.562500 GiB object_store_memory}\n",
      ". In total there are 0 pending tasks and 8 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.\n"
     ]
    },
    {
     "output_type": "error",
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-50-bedd9379228f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     66\u001b[0m         \u001b[0;31m#scheduler=ASHAScheduler())\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m         \u001b[0mscheduler\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mPopulationBasedTraining\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 68\u001b[0;31m         local_dir=\"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging\")\n\u001b[0m\u001b[1;32m     69\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Best hyperparameters found were: \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0manalysis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbest_config\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/tune.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(run_or_experiment, name, metric, mode, stop, time_budget_s, config, resources_per_trial, num_samples, local_dir, search_alg, scheduler, keep_checkpoints_num, checkpoint_score_attr, checkpoint_freq, checkpoint_at_end, verbose, progress_reporter, log_to_file, trial_name_creator, trial_dirname_creator, sync_config, export_formats, max_failures, fail_fast, restore, server_port, resume, queue_trials, reuse_actors, trial_executor, raise_on_failed_trial, callbacks, loggers, ray_auto_init, run_errored_only, global_checkpoint_period, with_server, upload_dir, sync_to_cloud, sync_to_driver, sync_on_checkpoint)\u001b[0m\n\u001b[1;32m    419\u001b[0m     \u001b[0mtune_start\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    420\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mrunner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_finished\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 421\u001b[0;31m         \u001b[0mrunner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    422\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mhas_verbosity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mVerbosity\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mV1_EXPERIMENT\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    423\u001b[0m             \u001b[0m_report_progress\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrunner\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprogress_reporter\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trial_runner.py\u001b[0m in \u001b[0;36mstep\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    400\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrial_executor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0min_staging_grace_period\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    401\u001b[0m                     \u001b[0mtimeout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0.1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 402\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_process_events\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# blocking\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    403\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrial_executor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_no_available_trials\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/trial_runner.py\u001b[0m in \u001b[0;36m_process_events\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    521\u001b[0m             \u001b[0;31m#  fetch_result functionality so that we don't timeout on fetch.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    522\u001b[0m             trial = self.trial_executor.get_next_available_trial(\n\u001b[0;32m--> 523\u001b[0;31m                 timeout=timeout)  # blocking\n\u001b[0m\u001b[1;32m    524\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mtrial\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    525\u001b[0m                 \u001b[0;32mreturn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py\u001b[0m in \u001b[0;36mget_next_available_trial\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    579\u001b[0m         \u001b[0;31m# See https://github.com/ray-project/ray/issues/4211 for details.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    580\u001b[0m         \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 581\u001b[0;31m         \u001b[0mready\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mshuffled_results\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    582\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mready\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    583\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/_private/client_mode_hook.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     45\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mclient_mode_enabled\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0m_client_hook_enabled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 47\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     48\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Projects/internal_projects/populationBasedTraining/pbt/lib/python3.7/site-packages/ray/worker.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(object_refs, num_returns, timeout, fetch_local)\u001b[0m\n\u001b[1;32m   1596\u001b[0m             \u001b[0mtimeout_milliseconds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1597\u001b[0m             \u001b[0mworker\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcurrent_task_id\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1598\u001b[0;31m             \u001b[0mfetch_local\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1599\u001b[0m         )\n\u001b[1;32m   1600\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mready_ids\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mremaining_ids\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpython/ray/_raylet.pyx\u001b[0m in \u001b[0;36mray._raylet.CoreWorker.wait\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpython/ray/_raylet.pyx\u001b[0m in \u001b[0;36mray._raylet.check_status\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import lightgbm as lgb\n",
    "import numpy as np\n",
    "import sklearn.datasets\n",
    "import sklearn.metrics\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from ray import tune\n",
    "from ray.tune.schedulers import ASHAScheduler\n",
    "\n",
    "\n",
    "def LightGBMCallback(env):\n",
    "    \"\"\"Assumes that `valid_0` is the target validation score.\"\"\"\n",
    "    _, metric, score, _ = env.evaluation_result_list[0]\n",
    "    tune.report(**{metric: score})\n",
    "\n",
    "\n",
    "def train_diabetes(config):\n",
    "    data = train_processed_df.drop('diabetes_mellitus', axis=1)\n",
    "    target = train_processed_df.diabetes_mellitus\n",
    "    #data, target = sklearn.datasets.load_breast_cancer(return_X_y=True)\n",
    "    train_x, test_x, train_y, test_y = train_test_split(\n",
    "        data, target, test_size=0.20)\n",
    "    train_set = lgb.Dataset(train_x, label=train_y)\n",
    "    test_set = lgb.Dataset(test_x, label=test_y)\n",
    "    gbm = lgb.train(\n",
    "        config,\n",
    "        train_set,\n",
    "        valid_sets=[test_set],\n",
    "        verbose_eval=False,\n",
    "        callbacks=[LightGBMCallback])\n",
    "    preds = gbm.predict(test_x)\n",
    "    pred_labels = np.rint(preds)\n",
    "    tune.report(\n",
    "        #mean_accuracy=sklearn.metrics.accuracy_score(test_y, pred_labels),\n",
    "        #binary_error=sklearn.metrics.accuracy_score(test_y, pred_labels),\n",
    "        binary_error=sklearn.metrics.roc_auc_score(test_y, pred_labels),\n",
    "        done=True)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    config = {\n",
    "        \"objective\": \"binary\",\n",
    "        \"metric\": \"binary_error\",\n",
    "        \"verbose\": -1,\n",
    "        #\"boosting_type\": tune.grid_search([\"gbdt\", \"dart\"]),\n",
    "        \"boosting_type\": \"gbdt\",\n",
    "        \"num_leaves\": tune.randint(10, 40),\n",
    "        'min_child_samples': tune.randint(100, 400),\n",
    "        'min_child_weight': tune.grid_search([1e-7,1e-6,1e-5, 1e-4, 1e-3, 1e-2,1e-1]),\n",
    "        'subsample': tune.uniform(0.4, 0.6),\n",
    "        'colsample_bytree': tune.uniform(0.2, 0.6),\n",
    "        'reg_alpha': tune.grid_search([0, 1, 2, 5, 7, 10, 50]),\n",
    "        'reg_lambda': tune.grid_search([0, 1, 5, 10, 20, 50]),\n",
    "        'scale_pos_weight': tune.grid_search([1, 2, 3]),\n",
    "        #\"learning_rate\": tune.loguniform(1e-2, 1e-1)\n",
    "        \"learning_rate\": 0.01\n",
    "    }\n",
    "\n",
    "    analysis = tune.run(\n",
    "        train_diabetes,\n",
    "        metric=\"binary_error\",\n",
    "        #mode=\"min\",\n",
    "        mode=\"max\",\n",
    "        resources_per_trial={\"cpu\": 7, \"gpu\": 0},\n",
    "        config=config,\n",
    "        num_samples=1,\n",
    "        #scheduler=ASHAScheduler())\n",
    "        scheduler=PopulationBasedTraining(),\n",
    "        local_dir=\"/Users/anitaclement/Documents/Projects/internal_projects/populationBasedTraining/logging\")\n",
    "\n",
    "    print(\"Best hyperparameters found were: \", analysis.best_config)"
   ]
  }
 ]
}